File: 2016-01/gates2.cpp
#include <bits/stdc++.h>
using namespace std;

int n;
string s;
vector<vector<int>> board(2000, vector<int>(2000, 0));
vector<vector<int>> visited(2000, vector<int>(2000, 0));

void flood(int i, int j) {
    vector<pair<int, int>> stack = {make_pair(i, j)};
    visited[i][j] = 1;
    while (stack.size()) {
        pair<int, int> p = stack.back();
        stack.pop_back();
        int dx[] = {0, 1, 0, -1};
        int dy[] = {1, 0, -1, 0};
        for (int d = 0; d < 4; d++) {
            int vX = (p.first + dx[d] + 2000) % 2000;
            int vY = (p.second + dy[d] + 2000) % 2000;
            if (board[vX][vY]) {
                continue;
            }
            if (!visited[vX][vY]) {
                stack.push_back(make_pair(vX, vY));
                visited[vX][vY] = 1;
            }
        }
    }
}

int main() {
    freopen("gates.in", "r", stdin);
    freopen("gates.out", "w", stdout);

    cin >> n;
    cin >> s;

    map<char, pair<int, int>> directionMap = {
        {'N', make_pair(0, 2)},
        {'S', make_pair(0, -2)},
        {'E', make_pair(2, 0)},
        {'W', make_pair(-2, 0)}
    };

    int curX = 0;
    int curY = 0;
    int prevX, prevY, betweenX, betweenY;
    for (int i = 0; i < n; i++) {
        prevX = curX; prevY = curY;
        curX = curX + directionMap[s[i]].first;
        curY = curY + directionMap[s[i]].second;
        betweenX = (prevX + curX) / 2;
        betweenY = (prevY + curY) / 2;
        board[(prevX + 2000) % 2000][(prevY + 2000) % 2000] = 1;
        board[(betweenX + 2000) % 2000][(betweenY + 2000) % 2000] = 1;
        board[(curX + 2000) % 2000][(curY + 2000) % 2000] = 1;
    }

    int components = 0;
    for (int i = 0; i < 2000; i++) {
        for (int j = 0; j < 2000; j++) {
            if (board[i][j]) {
                continue;
            }
            if (!visited[i][j]) {
                components++;
                flood(i, j);
            }
        }
    }
    cout << components - 1 << endl;
    return 0;
}

File: 2016-01/gates2.py
# surefire way to work
# oop, TLE
import sys
sys.stdin = open("gates.in")
sys.stdout = open("gates.out", 'w')
input = sys.stdin.readline

n = int(input())
s = input().strip()

directionMap = {
    "N": (0, 2), 
    "S": (0, -2), 
    "E": (2, 0), 
    "W": (-2, 0)
}

board = [[0 for _ in range(2000)] for _ in range(2000)]
curX, curY = 0, 0
for i in range(n):
    prevX, prevY = curX, curY
    curX += directionMap[s[i]][0]
    curY += directionMap[s[i]][1]
    betweenX, betweenY = (prevX + curX) // 2, (prevY + curY) // 2
    board[prevX][prevY] = 1
    board[betweenX][betweenY] = 1
    board[curX][curY] = 1

def flood(i, j):
    stack = [(i, j)]
    visited[i][j] = True
    while stack:
        uX, uY = stack.pop()
        for dx, dy in [(0, 1), (1, 0), (0, -1), (-1, 0)]:
            vX, vY = uX + dx, uY + dy
            if not (-2000 <= vX <= 1999 and -2000 <= vY <= 1999):
                continue
            if board[vX][vY]:
                continue
            if not visited[vX][vY]:
                stack.append((vX, vY))
                visited[vX][vY] = True

visited = [[False for _ in range(2000)] for _ in range(2000)]
components = 0
for i in range(2000):
    for j in range(2000):
        if board[i][j]:
            continue
        if not visited[i][j]:
            components += 1
            flood(i, j)
print(components - 1)

'''
abusing Python negative indexing
'''